### PR TITLE
Quark v2 changes

### PR DESCRIPTION
**Pull request by kevincheng96**
Updating the `quark-scripts` repo to depend on the latest version of Quark ([Quark v2](https://github.com/compound-finance/quark/pull/216)). The main changes in this PR are:

- Update `quark` submodule to the latest commit on the Quark v2 branch
- Bump up Solidity version to 0.8.27 and EVM version to `cancun`
- Update `QuarkOperation` struct to the newest version that has a `isReplayable` field
- Update `RecurringSwap.sol` to no longer need a `cancel` function and no longer need to call `allowReplay()`
- Bring in the `Cancel.sol` core script, used for cancelling replayable operations
- Update `QuarkBuilder` to use Quark v2; add the new fields `nonceSecret` and `totalPlays` to QuarkBuilder
- Optimize `QuarkBuilder` to get rid of stack too deep errors from introducing the new fields

### CODE REVIEW

File: src/builder/Actions.sol [hunk 1]
```diff
@@ -593,7 +593,8 @@ library Actions {
 
         // Construct QuarkOperation
         IQuarkWallet.QuarkOperation memory quarkOperation = IQuarkWallet.QuarkOperation({
-            nonce: accountState.quarkNextNonce,
+            nonce: bytes32(uint256(accountState.quarkNextNonce)),
```
Comment (kevincheng96) [inline src/builder/Actions.sol:596]: There's a bunch of interface changes that need to be made in `QuarkBuilder` for the client to pass in a random nonce. I plan to do that as a separate PR.

File: src/builder/PaycallWrapper.sol [hunk 2]
```diff
@@ -25,6 +25,7 @@ library PaycallWrapper {
 
         return IQuarkWallet.QuarkOperation({
             nonce: operation.nonce,
+            isReplayable: operation.isReplayable,
```
Comment (hayesgm) [inline src/builder/PaycallWrapper.sol:28]: This is interesting, and I think correct. The Paycall operation is repalyable IFF what it wraps is replayable?

File: src/builder/QuarkOperationHelper.sol [hunk 3]
```diff
@@ -101,6 +101,7 @@ library QuarkOperationHelper {
         IQuarkWallet.QuarkOperation memory lastQuarkOperation = quarkOperations[quarkOperations.length - 1];
         IQuarkWallet.QuarkOperation memory mergedQuarkOperation = IQuarkWallet.QuarkOperation({
             nonce: lastQuarkOperation.nonce,
+            isReplayable: lastQuarkOperation.isReplayable,
```
Comment (hayesgm) [inline src/builder/QuarkOperationHelper.sol:104]: Same here, so basically if it's `wrap` and `replayableSupply` this is now `replayableWrapAndSupply`?

**Review by hayesgm — APPROVED** [review]
lgtm

File: src/builder/PaycallWrapper.sol (hunk 2 above)
Comment (kevincheng96) [reply src/builder/PaycallWrapper.sol:28]: Exactly. If it's a `RecurringSwap` using `Paycall`, then the `Paycall` itself has to be replayable.

File: src/builder/QuarkOperationHelper.sol (hunk 3 above)
Comment (kevincheng96) [reply src/builder/QuarkOperationHelper.sol:104]: Yeah, if the last operation is replayable, the merged operation should be replayable.

**Comment by kevincheng96** [comment]
The latest commit (fc9d2094763fd27231b9770fdf7d3ba09e333509) rebases on main and applies the Quark v2 changes to the newest `MorphoClaimRewards` action. After the rebase, we were hitting stack too deep errors again, so the commit also includes some optimizations to the `QuarkBuilder` code.
